### PR TITLE
Us101 legal basis

### PR DESCRIPTION
**Pull request by madeye-matt**
- Added legalbasis as a separate entity 
- Added endpoint for getting all legal bases
- Added db migrations
- Added endpoint for creating surveys

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -121,7 +128,11 @@ func configureEnvironment() (dataSource, port string) {
 			dataSource = v
 		}
 
-		return dataSource, port
+		if v := os.Getenv("MIGRATION_SOURCE"); len(v) > 0 {
```
Comment (davidmort) [inline main.go]: is there a use case for migration source being configfurable? if's there's any db changes they should be in this repo and not come from anywhere else

File: models/surveys.go [hunk 2]
```diff
@@ -102,7 +132,75 @@ func NewAPI(db *sql.DB) (*API, error) {
...
+	if err == sql.ErrNoRows {
+		// Reference is unique - this is good
+		_, err = api.CreateSurveyStmt.Exec(surveyId, surveyRef, shortName, longName, legalBasis.Reference)
+
+		if err != nil {
+			http.Error(w, fmt.Sprintf("Create survey details failed - %v", err), http.StatusInternalServerError)
+			return
+		}
+
+		w.WriteHeader(http.StatusOK)
```
Comment (davidmort) [inline models/surveys.go]: this should return http.StatusCreated (201) along with the new resource that has been created

Comment (madeye-matt) [reply models/surveys.go]: Sure - copy and paste bug :-P

File: main.go (hunk 1 above)
Comment (madeye-matt) [reply main.go]: There is a case for this being configurable -  docker, cloud foundry etc will have the files in different places (main is installed in /usr/local/bin in docker for example).  I had trouble getting relative paths to work properly.  Now if there was a way to somehow package the files with the executable ...

File: db-migrations/5_legal_basis.up.sql [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+INSERT INTO survey.legalbasis ( ref, longname ) VALUES ( 'GovERD', 'GovERD' );
+INSERT INTO survey.legalbasis ( ref, longname ) VALUES ( 'STA1947', 'Statistics of Trade Act 1947' );
+INSERT INTO survey.legalbasis ( ref, longname ) VALUES ( 'STA1947_B', 'Statistics of Trade Act 1947 - BEIS' );
+INSERT INTO survey.legalbasis ( ref, longname ) VALUES ( 'Vol', 'Voluntary Not Stated' );
+INSERT INTO survey.legalbasis ( ref, longname ) VALUES ( 'Vol_B', 'Voluntary - BEIS' );
+
+UPDATE survey.survey SET legalbasis = 'GovERD' WHERE legalbasis = 'GovERD';
+UPDATE survey.survey SET legalbasis = 'STA1947' WHERE legalbasis = 'Statistics of Trade Act 1947';
+UPDATE survey.survey SET legalbasis = 'STA1947_B' WHERE legalbasis = 'Statistics of Trade Act 1947 - BEIS';
```
[the remark this answers is not in the document]
Comment (madeye-matt) [reply db-migrations/5_legal_basis.up.sql]: Picky but sure :-)

File: db-migrations/5_legal_basis.up.sql [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+UPDATE survey.survey SET legalbasis = 'Vol' WHERE legalbasis = 'Voluntary Not Stated';
+UPDATE survey.survey SET legalbasis = 'Vol_B' WHERE legalbasis = 'Voluntary - BEIS';
```
[the remark this answers is not in the document]
Comment (madeye-matt) [reply db-migrations/5_legal_basis.up.sql]: same as the comment above

File: models/surveys.go [hunk 5]
```diff
@@ -30,6 +30,12 @@ type Survey struct {
 	LongName   string `json:"longName"`
 	Reference  string `json:"surveyRef"`
 	LegalBasis string `json:"legalBasis"`
+	LegalBasisRef string `json:"legalBasisRef"`
```
[the remark this answers is not in the document]
Comment (madeye-matt) [reply models/surveys.go]: https://github.com/ONSdigital/rm-surveysvc-api/pull/6

**Comment by madeye-matt** [comment]
@davidmort @tej99 requested changes have been made apart from MIGRATION_SOURCE environment variable which we are going to leave in pending further investigation

**Comment by madeye-matt** [comment]
@tej99 I've added validation but the PR is now 1.7k files changed

**Comment by madeye-matt** [comment]
[US102: INT User to Create a CE - After R17/18 (M)](https://trello.com/c/9VgewZXM/77-us102-int-user-to-create-a-ce-after-r17-18-m)

**Comment by madeye-matt** [comment]
[US101: INT User to Create a Survey - After R17/18 (L)](https://trello.com/c/p7lh6cuO/68-us101-int-user-to-create-a-survey-after-r17-18-l)

**Comment by davidmort** [comment]
these changes look good to go, although there is an extra step needed for the jenkins build/deploy jobs to include the db-migrations folder in the built artifact

**Pull request merged**